### PR TITLE
feat(linter): eslint/sort-keys 

### PR DESCRIPTION
**Pull request by Goldziher**
This PR adds the `eslint/sort-keys` implementation as part of #479 

Note: two tests are commented out. I was not able to model this behavior properly. I would be happy to have some assistance there.

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:185]: Use a diagnostic factory function. See other rules for examples.

**Comment by DonIsaac** [comment]
Thanks for your contribution to oxc! This looks pretty good, just a few things to polish off.

File: .gitignore
Comment (Goldziher) [reply .gitignore:31]: Ok, why?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:62]: Ok. Which one would this rule be? Style?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:72]: 👍🏼

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:185]: 👍🏼

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:76]: 👍🏼

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:62]: im proceeding for now with sytle. Lemme know if something else is better.

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:54]: 👍🏼

File: .gitignore
Comment (Goldziher) [reply .gitignore:31]: done. 

I simply don't commit it - but I do commit it to other projects, which are not open source, so I am not globally ignoring it. I would recommend its ignored in OXC instead.

**Comment by Goldziher** [comment]
> Thanks for your contribution to oxc! This looks pretty good, just a few things to polish off.

Great. I addressed all comments.

File: Cargo.toml
Comment (DonIsaac) [inline Cargo.toml:195]: Please undo this

File: .gitignore
Comment (rzvxa) [reply .gitignore]: Feel free to mark this conversation as resolved, We can add it in a follow-up if Boshen gives a 👍

@DonIsaac

File: Cargo.toml
Comment (Goldziher) [reply Cargo.toml:195]:
```suggestion
```

Comment (Goldziher) [reply Cargo.toml]: unintentional. Reverted. Note- this is an issue for running a debugger in intellij.

Comment (DonIsaac) [reply Cargo.toml]: Yup, same here for me in VSCode. I simply don't commit it; it helps keep CI nice and speedy 🏃

File: .gitignore
Comment (DonIsaac) [reply .gitignore]: Boshen has expressed that editor-specific files should not be included in oxc's .gitignore. I don't recall his reasoning, but I know he doesn't want it there.

**Review by DonIsaac — APPROVED** [review]

**Comment by DonIsaac** [comment]
@rzvxa feel free to merge this when CI is passing and you are ready.

**Comment by rzvxa** [comment]
@Goldziher It seems like you've done the hard part, Is there anything that I can help with to push this PR past the finish line?

To fix the CI issues you can run `just r` on your local machine.
For updating snapshots you should do the following:

1. `cargo insta test`
2. `cargo insta review`
3. accept or reject snapshot updates
3. commit the updated snapshots.

The snapshot you've committed is an intermediate one. `Insta` on tests creates `new` snapshots and after review, it would remove it in favor of the "actual" snapshot.

The one with `new` in the name is for local use in the review process and shouldn't be committed.

**Comment by Goldziher** [comment]
> @Goldziher It seems like you've done the hard part, Is there anything that I can help with to push this PR past the finish line?
> 
> To fix the CI issues you can run `just r` on your local machine. For updating snapshots you should do the following:
> 
> 1. `cargo insta test`
> 2. `cargo insta review`
> 3. accept or reject snapshot updates
> 4. commit the updated snapshots.
> 
> The snapshot you've committed is an intermediate one. `Insta` on tests creates `new` snapshots and after review, it would remove it in favor of the "actual" snapshot.
> 
> The one with `new` in the name is for local use in the review process and shouldn't be committed.

thanks, its done.

**Comment by rzvxa** [comment]
There are still failing CI tasks, Have you tried using `cargo lint -- -D warnings` or `just r` commands?

**Comment by Goldziher** [comment]
> There are still some failing CI tasks, Have you tried using `cargo lint -- -D warnings` or `just r` commands?

I'll try running it tomorrow, cheers

**Comment by Goldziher** [comment]
Hi, @rzvxa - i fixed all the clippy issues and ensured the tests are all passing. I am getting one error running `just r` that I don't quite understand:

```text
est linter::rules::test::test_doc_pages ... FAILED
test linter::rules::test::test_rules_table ... ok

failures:

---- linter::rules::test::test_doc_pages stdout ----
  x Expected a semicolon or an implicit semicolon after a statement, but found
  | none
   ,-[Pedantic/eslint/sort-keys/code-snippet:1:10]
 1 | let myObj {
   :          ^
 2 |   c: 1,
   `----
  help: Try insert a semicolon here

thread 'linter::rules::test::test_doc_pages' panicked at tasks/website/src/linter/rules/test.rs:94:64:
called `Result::unwrap()` on an `Err` value: "1 errors occurred while parsing Pedantic/eslint/sort-keys/code-snippet.jsx"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    linter::rules::test::test_doc_pages

test result: FAILED. 5 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.16s

error: test failed, to rerun pass `-p website --lib`
error: Recipe `test` failed on line 67 with exit code 101
error: Recipe `ready` failed on line 27 with exit code 101
```

I don't understand what semi-colon it wants from me to add - the JS is find and there is a semi-colon at the end of the statement.

**Comment by rzvxa** [comment]
You are missing an equal sign there:

```js
let obj = { /* ... */ }
```
Since there is no equal sign it interprets it as two expressions e.g.
```js
let obj; { /* ... */ }
```
And says to add a semicolon to be more explicit.

**Comment by Goldziher** [comment]
> You are missing an equal sign there:
> 
> ```js
> let obj = { /* ... */ }
> ```
> 
> Since there is no equal sign it interprets it as two expressions e.g.
> 
> ```js
> let obj; { /* ... */ }
> ```
> 
> And says to add a semicolon to be more explicit.

🤦🏼 

All fixed and everything is green locally. Thanks for your help!

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (jelly) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:89]: The TODO comments can be removed as they placeholder has already been replaced.

Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:89]: fixed

**Comment by Goldziher** [comment]
@rzvxa is this PR done?

**Comment by DonIsaac** [comment]
@rzvxa appears busy, I'll review this in the morning.

**Comment by Goldziher** [comment]
> @rzvxa appears busy, I'll review this in the morning. 

Anything?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:91]: Consider `Self::default()` for clarity

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:215]: Can you zip these two iterations together without collecting?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:251]: Can we not return `Ordering::Equal` here?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:341]: What happened here?

**Comment by DonIsaac** [comment]
Apologies for the delay, I got sidetracked 😅

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:341]: see the PR description - there are two cases where I wasn't able to match the test cases.

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:251]: I am not 100% sure, thats why I kept it like this.

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:215]: done

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:91]: done

**Comment by Goldziher** [comment]
@DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.

**Comment by DonIsaac** [comment]
> @DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.

The conflicts aren't too bad, I believe in you 😄

**Comment by Goldziher** [comment]
> > @DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.
> 
> The conflicts aren't too bad, I believe in you 😄 

That's not the issue. Ill need to add a second remote now, but I'll do it.

**Comment by Goldziher** [comment]
Should be good to go @DonIsaac

**Comment by DonIsaac** [comment]
Once CI is passing we can merge this

**Pull request merged**
